content/en/blog/_posts/2024-11-11-dynamic-kubernetes-api-server-for-cozystack/index.md

```diff
@@ -0,0 +1,454 @@
+All the rest logic is handled by FluxCD. This pattern allows us to easily extend the platform with new
+applications and provide the ability to create new applications that just need to be packaged
+into the appropriate Helm chart.
+
+{{< figure src="cozystack.png" alt="Interface of the Cozystack platform" caption="Interface of the Cozystack platform" >}}
+
+So, in our platform, everything is configured as `HelmRelease` resources. However, we ran into
+two problems: limitations of the RBAC model and the need for a public API. Let's delve into these
+
+### Limitations of the RBAC model
+
+The standard RBAC system in Kubernetes doesn't allow you to restrict access to a list of resources
+of the same kind based on labels or specific fields in the spec. When creating a role, you can limit
+access across the resources in the same kind only by specifying specific resource names in `resourceNames`.
+For methods like GET, PATCH, or UPDATE it will work. However, filtering by `resourceNames` in the LIST
+method doesn't work like that. Thus you can limit listing certain resources by kind but not by name.
+
+So, we decided to introduce new resource types based on the names of the Helm charts they use and
+generate the list of available kinds dynamically at runtime in our extension api-server.
+This way, we can reuse Kubernetes standard RBAC model to manage access to specific resource types.
+
+### Need for a public API
+
+Since our platform provides capabilities for deploying various managed services, we want to organize
+public access to the platform's API. However, we can't allow users to interact directly with resources
+like `HelmRelease` because that would let them specify arbitrary names and parameters for Helm charts to
+deploy, potentially compromising our system.
+
+We wanted to give users the ability to deploy a specific service simply by creating the resource with corresponding
+kind in Kubernetes. The type of this resource should be named the same as the chart from
+which it's deployed. Here are some examples:
+
+- `Kind: Kubernetes` → `chart: kubernetes`
+- `Kind: Postgres` → `chart: postgres`
+- `Kind: Redis` → `chart: redis`
+- `Kind: VirtualMachine` → `chart: virtual-machine`
+
+Moreover, we don't want to have to add a new type to codegen and recompile our extension api-server
+every time we add a new chart for it to start being served.
+The update should be done dynamicly or provided via a `ConfigMap` by the administrator.
+
+### Two-Way conversion
+
+Currently, we already have integrations and a dashboard that continue to use `HelmRelease` resources.
+At this stage, we didn't want to lose the ability to support this API. Considering that we're simply
+translating one resource into another, support is maintained and it works both ways.
+If you create a `HelmRelease`, you'll get a custom resource in Kubernetes, and if you create a
+custom resource in Kubernetes, it will also be available as a `HelmRelease`.
```

Comment:
```suggestion
Currently, we already have integrations and a dashboard that continue to use HelmRelease resources.
At this stage, we didn't want to lose the ability to support this API. Considering that we're simply
translating one resource into another, support is maintained and it works both ways.
If you create a HelmRelease, you'll get a custom resource in Kubernetes, and if you create a
custom resource in Kubernetes, it will also be available as a HelmRelease.
```